### PR TITLE
REGR: Fix bug when replacing categorical value with self

### LINKED ISSUE #33288: REGR: Replacing a category with itself replaces it with np.nan
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] I have confirmed this bug exists on the master branch of pandas.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
pd.Series(["a", "b"]).astype("category").replace("a", "a")
# 0    NaN
# 1      b
# dtype: category
# Categories (1, object): [b]
```

Operating on the categorical array directly, i.e. `pd.Categorical(["a", "b"]).replace("a", "a")` yields the same result.

#### Problem description

Replacing a category with itself replaces it with `np.nan`. This problem was introduced with 1.0.0.

#### Expected Output

I would have expected the behavior from 0.25.3:
```python
pd.Series(["a", "b"]).astype("category").replace("a", "a")
# 0    a
# 1    b
# dtype: category
# Categories (2, object): [a, b]
```

Note that if we work with lists, we get
```python
pd.Series(["a", "b"]).astype("category").replace(["a"], ["a"])
# dtype: category
# 0    a
# 1    b
# type: object
```
which is also not what I would expect, because we're now losing the dtype. This behavior has been described elsewhere (e.g. https://github.com/pandas-dev/pandas/pull/31734#issuecomment-583130043) and it's consistent with 0.25.3.

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.2.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.4.0-176-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200325
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
```

</details>

### PR DESCRIPTION
**Pull request by dsaxton**
- [ ] closes #33288
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
LGTM pending green

**Comment by MomIsBestFriend** [comment]
@dsaxton If you rebase it should fix the Travis issue

**Comment by simonjayhawkins** [comment]
@dsaxton I've tagged this as 1.0.4 for now since #33288 is a regression.

I don't think need to move whatsnew yet, see https://github.com/pandas-dev/pandas/pull/33157#issuecomment-606742451

opened #33300 to help keep track and discuss

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -2463,7 +2463,8 @@ def replace(self, to_replace, value, inplace: bool = False):
                 if new_value in cat.categories:
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:2463]: you can move this check to line 2451, e.g.
```
elif replace_value == new_value:
    continue
```

File: pandas/tests/arrays/categorical/test_algos.py [hunk 2]
```diff
@@ -64,6 +64,7 @@ def test_isin_cats():
     [
         ("b", "c", ["a", "c"], "Categorical.categories are different"),
         ("c", "d", ["a", "b"], None),
+        ("a", "a", ["a", "b"], None),
```
Comment (jreback) [inline pandas/tests/arrays/categorical/test_algos.py:67]: can you annotate this case (add a comment on line before with this issue number)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dsaxton

**Pull request merged**